### ISSUE TITLE
Allow an array of strings in <Route path>

### ISSUE THREAD

**Issue by mjackson** [ReactTraining/react-router#5866]
This is a follow-up from #5393.

It would be nice if `<Route path>` could accept an array of strings instead of a single string. This would make it possible to render the same `<Route>` element at multiple different URLs [as suggested here](https://github.com/ReactTraining/react-router/pull/5393#issuecomment-320862212).

The reason #5393 was not merged is because there was [a problem with the way the code worked with respect to relative routes](https://github.com/ReactTraining/react-router/pull/5393#issuecomment-357150473). When this work is done, `match.path` should still be a string, not an array.

**Comment by o-mega** [comment]
It works but not as expected.
There is no `props.match.params` data for each path in an array.

**Comment by crobinson42** [comment]
Working great, just throwing the propTypes warning mentioned above 👍

**Comment by kyleramirez** [comment]
What needs to be done to remove this prop-types warning?

**Comment by dorrogeray** [comment]
Hi,

there seems to be an issue with match.params when defining multiple paths for one route with same named params.

Example, using following paths:
```javascript
const paths = [
	'/:mutation(cs)',
	'/:mutation(cs)-:currency(czk|eur)',
]
```
Url **/cs** will have match.params.mutation **undefined**. It clearly gets overwritten by the 'non matched' mutation from the second path, which is not matched.

This is likely comes from the way pathToRegexp works.

```javascript
let keys = [];
pathToRegexp(paths, keys);
```
Will set keys to:
```json
[
  {
    "name": "mutation",
    "prefix": "/",
    "delimiter": "/",
    "optional": false,
    "repeat": false,
    "partial": false,
    "asterisk": false,
    "pattern": "cs"
  },
  {
    "name": "mutation",
    "prefix": "/",
    "delimiter": "/",
    "optional": false,
    "repeat": false,
    "partial": true,
    "asterisk": false,
    "pattern": "cs"
  },
  {
    "name": "currency",
    "prefix": "",
    "delimiter": "/",
    "optional": false,
    "repeat": false,
    "partial": false,
    "asterisk": false,
    "pattern": "czk|eur"
  }
]
```

The workaround is to name the param in first path differently, and then resolve the params manually.

```javascript
const paths = [
	'/:mutationA(cs)',
	'/:mutationB(cs)-:currency(czk|eur)',
]
```

Using version 4.2.2.

**Comment by maxparelius** [comment]
Getting same issue as @dorrogeray in one of my projects. The workaround to have different named params for each route is not great. What is the status on this PR? This is highly needed and makes sense to have as it aligns with pathToRegexp expected behavior. When will this go live?

**Comment by acro5piano** [comment]
It would be really happy if this feature will be added, but for now I found a solution for my use case.

I made a parameter optional, for use cases like:

- Route has common path such as `/users` and `/users/create`
- Want to mount once for those compnents

before:

```
<Route exact path="/users" component={UserList} />
<Route exact path="/users/create" component={UserList>} />
```

after: 

```
<Route exact path="/users/:action?" component={UserList>} />
```

`before` renders twice while `after` renders once.

Furthermore, `UserList.props` should be like this:

```
{
    match: {
        params: {
            action: undefined | string
        }
    }
}
```

refs: 

- https://github.com/ReactTraining/react-router/issues/4551#issuecomment-281454148
- https://www.npmjs.com/package/path-to-regexp

**Comment by iamdanthedev** [comment]
```
declare module "react-router" {
  import { RouteProps as RoutePropsOrig } from "@types/react-router";
  import * as React from "react";
  export * from "@types/react-router";

  export type RouteProps = Omit<RoutePropsOrig, "path"> & {
    path: string[] | string;
  };

  export class Route extends React.Component<RouteProps, any> {}
}
```

```
// @ts-ignore
Route.propTypes.path = any;
```

solve the problem :-)

**Comment by mjackson** [comment]
Closed by #5889.
